### ISSUE TITLE
Support nil interfaces

### ISSUE THREAD

**Issue by wade-tattersall** [liip/sheriff#19]
Marshal currently triggers a panic if it runs into any nil pointers. 

I think such values should be stored in the resulting map as nils. This would be consistent with the json package, allowing the resulting map to be passed to json.Marshal and get the same result as if the struct itself were passed directly to json.Marshal.

Example code:
```package main

import (
	"encoding/json"
	"fmt"
	"github.com/liip/sheriff"
)

func main() {
	expected := map[string]interface{}{"thing": nil}
	fmt.Println(expected)

	type Example struct {
		Thing *int `json:"thing" groups:"include"`
	}
	x := Example{}
	_, _ = json.Marshal(x) // {"thing":null}

	opt := sheriff.Options{
		Groups: []string{"include"},
	}
	result2, err := sheriff.Marshal(&opt, x)
	if err != nil {panic(err)}
	fmt.Println(result2)
}```

**Comment by mweibel** [comment]
@wade-tattersall 
Thanks for the repro case. However I couldn't find any panic. Could you elaborate which panic you get?
Your example seems to work fine:
https://play.golang.org/p/fyoe2I_QG-r

**Comment by wade-tattersall** [comment]
Sorry, I messed up somehow - I think I must have forgotten to save before running the repro after I tried simplifying it.

After a bit of digging, I was able to work out what does trigger it: https://play.golang.org/p/7wCzhabD1dF.

That is, it fails for the relatively trivial case of `sheriff.Marshal(&opt, nil)`, and hence also fails when `marshalValue` calls `Marshal` for any member field that is a nil interface. 

There's a subtlety there that confuses me - `marshalValue` only calls `Marshal` for `k == reflect.Interface || k == reflect.Struct`, but when it does so, the resulting parameter that Marshal receives has a `ValueOf(data).Kind() == reflect.Invalid`. So it seems to lose the type information as it goes.

**Comment by mweibel** [comment]
I don't have much time at the moment - if you provide a PR though (with test cases please), I'll do my best to review it!

**Comment by xuancanh** [comment]
Hi @mweibel, I've created a PR for this issue. Could you help review it? Thank you a lot.

**Comment by mweibel** [comment]
fixed by #26
